src/cluster_migrate.c

```diff
@@ -0,0 +1,2003 @@
+void initSlotExportJobClient(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_EXPORT);
+    job->client = createClient(job->conn);
+    job->conn = NULL;
+    job->client->flag.authenticated = 1;
+    job->client->slot_migration_job = job;
+    initClientReplicationData(job->client);
+}
+
+/* Generate and store the SYNCSLOTS ESTABLISH command to send to the target for the job. */
+sds generateSyncSlotsEstablishCommand(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_EXPORT);
+    sds result = sdscatprintf(sdsempty(),
+                              "*%ld\r\n$7\r\nCLUSTER\r\n$9\r\nSYNCSLOTS\r\n"
+                              "$9\r\nESTABLISH\r\n$6\r\nSOURCE\r\n$40\r\n"
+                              "%.40s\r\n$4\r\nNAME\r\n$40\r\n%.40s\r\n"
+                              "$10\r\nSLOTSRANGE\r\n",
+                              8 + listLength(job->slot_ranges) * 2,
+                              server.cluster->myself->name,
+                              job->name);
+    listIter li;
+    listNode *ln;
+    listRewind(job->slot_ranges, &li);
+    while ((ln = listNext(&li))) {
+        slotRange *range = (slotRange *)ln->value;
+        sdscatfmt(result, "$%i\r\n%i\r\n$%i\r\n%i\r\n",
+                  digits10(range->start_slot), range->start_slot,
+                  digits10(range->end_slot), range->end_slot);
+    }
+    return result;
+}
+
+/* There are two potential triggers for streaming (whichever happens first):
+ *   1. SYNCSLOTS REQUEST-PAUSE command
+ *   2. BGSAVE child process completes
+ */
+void slotExportBeginStreaming(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_EXPORT);
+    updateSlotMigrationJobState(job, SLOT_EXPORT_STREAMING);
+
+    /* When the slot export is not ready, it will skip adding the client to the
+     * pending write queue (creating a backlog of pending commands). If any
+     * data is pending there, we need to manually put it in the write queue to
+     * flush it. */
+    putClientInPendingWriteQueue(job->client);
+
+    serverLog(LL_NOTICE,
+              "Slot migration %s snapshot finished, starting streaming.", job->description);
+}
+
+/* Attempt to pause the provided slot export job. If we can pause, the state will be updated to
+ * SLOT_EXPORT_FAILOVER_PAUSED and the PAUSED subcommand is sent to the target. Otherwise, there is
+ * no effect. */
+void slotExportTryDoPause(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_EXPORT);
+
+    if (server.debug_slot_migration_prevent_pause ||
+        (server.slot_migration_max_failover_repl_bytes >= 0 &&
+         getClientOutputBufferMemoryUsage(job->client) > (size_t)server.slot_migration_max_failover_repl_bytes)) {
+        return;
+    }
+    serverLog(LL_NOTICE,
+              "Pausing writes to allow slot migration %s to finalize failover.",
+              job->description);
+    job->mf_end = mstime() + server.cluster_mf_timeout * CLUSTER_MF_PAUSE_MULT;
```

Comment:
if we are reusing cluster_mf_timeout, maybe we should also mention it in the valkey.conf?

or we won't (untill we add the two steps process)